presto-main/src/main/java/com/facebook/presto/operator/WindowOperator.java

```diff
@@ -484,12 +484,51 @@ static int findEndPosition(int startPosition, int endPosition, BiFunction<Intege
         checkArgument(endPosition > 0, "endPosition must be greater than zero: %d", endPosition);
         checkArgument(startPosition < endPosition, "startPosition must be less than endPosition: %d < %d", startPosition, endPosition);
 
-        // TODO: do position binary search
-        for (int first = startPosition, second = startPosition + 1; second < endPosition; first++, second++) {
-            if (!comparator.apply(first, second)) {
-                return second;
+        int left = startPosition;
+        int right = endPosition - 1;
+        for (int i = 0; i < endPosition - startPosition; i++) {
+            int distance = right - left;
+
+            if (distance == 0) {
+                return right + 1;
+            }
+
+            if (distance == 1) {
+                if (comparator.apply(left, right)) {
+                    return right + 1;
+                }
+                else {
+                    return right;
+                }
+            }
+
+            int mid = left + distance / 2;
+            if (comparator.apply(left, mid)) {
+                // explore to the right
+                left = mid;
+            }
+            else {
+                // explore to the left
+                right = mid;
+            }
+        }
+
+        // hasn't managed to find a solution after N iteration. Probably the input is not sorted. Lets verify it.
+        for (int first = startPosition; first < endPosition; first++) {
+            boolean previousPairsWereEqual = true;
+            for (int second = first + 1; second < endPosition; second++) {
```

Comment:
This sanity check should never happen (because data has to be sorted, otherwise WFs will fail in dozen of other places), so it's probably okay to not complicate this code more by performing more sophisticated pass here.